### ISSUE TITLE
Make it possible to mark events with vertical lines

### ISSUE THREAD

**Issue by ssm** [munin-monitoring/munin#563]
Munin should be able to mark graphs with events (like "new code deployed") to a system, and having that visible on several graphs.

The relevant RRD feature is VRULE, which draws a vertical line at a specific time. (with optional colour and legend)

This should be applied to one or more group, host or service FQNs.

An "event datafile" sqlite db could be a simple way to add this.  Anything with write access could write to it, and munin could read from it to add vrules where FQN matches.

Example event:

    ---
    time: 1442242424
    legend: v1.1.4 deployed
    colour: pink
    target:
      example.com;frontend
      example.net;backend;db1.example.com:postgresql

**Comment by chteuchteu** [comment]
We could handle this using two ways:
- a command line tool to list, add and edit events
- a page on the web UI with the same features. We would interact with the event datafile using munin's API using Ajax

**Comment by ssm** [comment]
There are multiple parts to this.

# The basics
* How and where to store events (persistently)
* How long to keep events, and what should be responsible for expiry.
* How to add and remove events manually, without going through munin.

The easiest would be to store the events using DBI and SQLite to store them.

Keep events for a configurable maximum time (easy), or looking at how long the RRDs actually keep data to expire useless events (harder).

Document the SQL needed to do manual updating (easy), or make a maintenance utility to do the work, assuming exclusive database access (endless source of feature creep).

# Internal API
* Make a perl module responsible for events, giving it the requisite methods to:
  * Add an event
  * Remove an event
  * Search for events
  * Expire old events

This is not too much work.  A "crud" kind of module, only without a web interface. Yet.

# External API
These should both use the internal API, making them subject to the same restrictions, and use the same data validation.
* Make a web endpoint to map the internal API
* Make a command line utility to submit events

**Comment by chteuchteu** [comment]
Great comments!
Should we allow editing an event?
What's the milestone for all of this?

The web interface would be quite straightforward: a list of events sorted by date, with simple actions for each (edit / delete), and a form to add a new one.
Data would be saved and retrieved using Ajax, by calling the web endpoint.

We should also update the JSON/XML "API" (apache-day.json) to send those events altogether with graph data.

**Comment by ssm** [comment]
Someone will, at some point, end up with multiple gigabytes of events.  This should be considered.

**Comment by chteuchteu** [comment]
Yes indeed, even if I personally doubt it will (if we come to a way to delete old events, and taking in account the fact that's it's just text)
About the web UI: we may limit the amount of events we retrieve
And for the JSON/XML API: we will only send event according to the appropriate time range (day|week|month|year)

**Comment by steveschnepp** [comment]
Some time ago, I was planning a `.alert` field attribute to be able for the plugin to declare events. It doesn't really fit exactly the same need, but it fits well in Munin's way of doing things : plugin initiated.

**Comment by chteuchteu** [comment]
That's off-topic, but it would actually be awesome to have a list of previous alerts (warning / critical).
We'd keep track of all alerts, and list those in an independing page, plus on each plugin page having an alert
And have the ability to remove one (if it isn't important)

Those events ("from" date, duration / "to" date, plugin & host, criticality) wouldn't manually be triggered by plugins, but handled by munin's alert engine.

**Comment by shapirus** [comment]
Take a look at the TICK draw type (https://oss.oetiker.ch/rrdtool/doc/rrdgraph_graph.en.html).
Looks like it's intended for the exact purpose that we're trying to achieve here. VRULE doesn't seem to me relevant for this purpose.

**Comment by ssm** [comment]
yes, the TICK draw type seems to be the correct one.

**Comment by thebeline** [comment]
Has there been any progress on this?  I would love to see this feature come to Munin...  It would make things like notifying of a launch, or notifying of particular process start or end times, awesome...

**Comment by ssm** [comment]
Related feature request: #720
